### ISSUE TITLE
Unwanted hyphen appears between Chinese character and Latin word

### ISSUE THREAD

**Issue by yuanshuai0217** [koreader/koreader#11478]
* KOReader version: 2024.01
* Device: moaan mix7 (with Android 11)

#### Issue

Under some circumstance, unwanted hyphen will appear between Chinese character and Latin word.

#### Steps to reproduce

This is a MWE file: [test.html.zip](https://github.com/koreader/koreader/files/14322851/test.html.zip)

Under my preset (Font size=22.0, Typography rules=Simplified Chinese; Hyphenation=English(US), Hyphenation dictionary=English(US)), an unwanted hyphen will appears here:

![Reader_2024-02-18_221830](https://github.com/koreader/koreader/assets/31814374/c0896c6a-30fc-4a5d-b870-960484e2171b)

**Comment by Frenzie** [comment]
Is it always unwanted after a Chinese character or is it something specific to this example?

**Comment by poire-z** [comment]
It is unwanted (you can trust my cr-intimate knowledge of Chinese & JK :)).
This is an issue, I can reproduce, investigating.

**Comment by poire-z** [comment]
A bit hard to get again into the "find a better wrap opportunity at end of line if there's too much spacing by looking at hyphenation opportunities in the overflowing next word" code (to the point that it felt dubious :) but it works, so not touching it).
I think this issue is just that we were giving to the "find hyphenation opportunities in this word" code a word made of CJK chars + latin chars (in this case: `是COVID`), and it somehow tells that we can between the CJK char and the next latin char, so this CJK char ends up flagged as "draw hyphen after".
It can be solved by just properly not including CJK chars in "words" candidate for hyphenation:
```diff
--- a/crengine/src/lvstring.cpp
+++ b/crengine/src/lvstring.cpp
@@ -5078,3 +5078,3 @@ void lStr_findWordBounds( const lChar32 * str, int sz, int pos, int & start, int
         lUInt16 props = getCharProp(str[cur]);
-        if ( props & (CH_PROP_ALPHA|CH_PROP_HYPHEN) )
+        if ( props & (CH_PROP_ALPHA|CH_PROP_HYPHEN) && !lStr_isCJK(str[cur]) )
             break;
@@ -5093,3 +5093,3 @@ void lStr_findWordBounds( const lChar32 * str, int sz, int pos, int & start, int
         lUInt16 props = getCharProp(ch);
-        if ( props & (CH_PROP_ALPHA|CH_PROP_HYPHEN) ) {
+        if ( props & (CH_PROP_ALPHA|CH_PROP_HYPHEN) && !lStr_isCJK(ch) ) {
             first = cur;
@@ -5111,3 +5111,3 @@ void lStr_findWordBounds( const lChar32 * str, int sz, int pos, int & start, int
         lUInt16 props = getCharProp(str[cur]);
-        if ( !(props & (CH_PROP_ALPHA|CH_PROP_HYPHEN|CH_PROP_MODIFIER)) )
+        if ( !(props & (CH_PROP_ALPHA|CH_PROP_HYPHEN|CH_PROP_MODIFIER)) || lStr_isCJK(str[cur]) )
             break;
```
(yes, ugly wording, we tag as CH_PROP_ALPHA any letter, even CJK "letters"/non-punctuation).
This looks like it is enough to fix this issue, hoping it doesn't cause others, you'll tell me :)

Mentionning another issue I noticed while testing, with screenshots for reference:
I have added after that chinese full-stop (that little circle): a space, and then "Ugxwnted..." (originally "Unwanted...", twisted to prevent hyphenation `Un-wanted`) - so no line break, no new paragraph - and got:
![image](https://github.com/koreader/koreader/assets/24273478/4acc975e-79e3-4fd6-acbe-1dffa7174229)
I think this blank at end of line shouldn't be here, and the full-stop should be half-width and should touch the right margin, like this:
![image](https://github.com/koreader/koreader/assets/24273478/a961371c-5169-431f-8245-4b4215b2c409)

I can have that fixed with, early in `AddLine(start, end,...)`:
```diff
         // Note: not certain why or how useful this lastnonspace (used below) is.
         int lastnonspace = 0;
         for ( int k=end-1; k>=start; k-- ) {
             // Also not certain if we should skip floats or LCHAR_IS_OBJECT
             if ( !(m_flags[k] & LCHAR_IS_SPACE) ) {
                 lastnonspace = k;
                 break;
             }
         }
+        // Handle some edge case here (can't find another place where we could
+        // handle it): when this line ends with a CJK char and then a space.
+        // Usually, this trailing space is made part of the last word (and this
+        // last word gets its width reduced so this space is like it was not there).
+        // When making CJK words, we make a word on each CJK char, not knowing
+        // yet what comes after, and we would end with a space-only word, that
+        // would get a width of 0, but would prevent any CJK flexible width at
+        // end-of-line tuning for that previous CJK char/word, as it is not at
+        // end of line because of this space char...
+        if ( lastnonspace < end-1 && lastnonspace >= start && (m_flags[lastnonspace] & LCHAR_IS_CJK) ) {
+            end = lastnonspace+1; // ignore last space(s)
+        }
```
(at least, now I can be _certain why and how useful this lastnonspace is_ :))

This fix will probably land into KOReader with other stuff only after next stable release, next month.

**Comment by yuanshuai0217** [comment]
I think the problem has been solved. Thanks for your replying.

**Comment by poire-z** [comment]
Let's keep this issue opened until it's actually fixed in the current KOReader people can install :) (next month)
